src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,529 @@
+            }
+
+            writer.Write(propertyName, filePath);
+        }
+
+        internal static string EncodeByteArrayValue(ReadOnlySpan<byte> value)
+        {
+            var builder = PooledStringBuilder.GetInstance();
+            EncodeByteArrayValue(value, builder.Builder);
+            return builder.ToStringAndFree();
+        }
+
+        internal static void EncodeByteArrayValue(ReadOnlySpan<byte> value, StringBuilder builder)
+        {
+            foreach (var b in value)
+            {
+                builder.Append(b.ToString("x"));
+            }
+        }
+
+        protected void WriteByteArrayValue(JsonWriter writer, string name, ReadOnlySpan<byte> value)
+        {
+            var builder = PooledStringBuilder.GetInstance();
+            EncodeByteArrayValue(value, builder.Builder);
+            writer.Write(name, builder.ToStringAndFree());
+        }
+
+        protected void WriteType(JsonWriter writer, string key, Type? type)
+        {
+            writer.WriteKey(key);
+            WriteType(writer, type);
+        }
+
+        protected void WriteType(JsonWriter writer, Type? type)
+        {
+            if (type is null)
+            {
+                writer.WriteNull();
+                return;
+            }
+
+            writer.WriteObjectStart();
+            writer.Write("fullName", type.FullName);
+            // Note that the file path to the assembly is deliberately not included here. The file path
+            // of the assembly does not contribute to the output of the program.
+            writer.Write("assemblyName", type.Assembly.FullName);
+            writer.Write("mvid", type.Assembly.ManifestModule.ModuleVersionId.ToString());
+            writer.WriteObjectEnd();
+        }
+
+        private (JsonWriter, PooledStringBuilder) CreateWriter()
+        {
+            var builder = PooledStringBuilder.GetInstance();
+            var writer = new StringWriter(builder);
+            return (new JsonWriter(writer), builder);
+        }
+
+        internal string GetKey(
+            CompilationOptions compilationOptions,
+            ImmutableArray<SyntaxTreeKey> syntaxTrees,
+            ImmutableArray<MetadataReference> references,
+            ImmutableArray<byte> publicKey,
+            ImmutableArray<AdditionalText> additionalTexts,
+            ImmutableArray<DiagnosticAnalyzer> analyzers,
+            ImmutableArray<ISourceGenerator> generators,
+            ImmutableArray<KeyValuePair<string, string>> pathMap,
+            EmitOptions? emitOptions,
+            DeterministicKeyOptions options,
+            CancellationToken cancellationToken)
+        {
+            additionalTexts = additionalTexts.NullToEmpty();
+            analyzers = analyzers.NullToEmpty();
+            generators = generators.NullToEmpty();
+
+            var (writer, builder) = CreateWriter();
+
+            writer.WriteObjectStart();
+
+            writer.WriteKey("compilation");
```

Comment:
nit: consider moving all these .WriteKey's into the respective WriteXXX helpers.

Reply:
Originally I had this but pulled into this form because several `WriteXXX` helpers are used with different key values. Most notably is actually `WriteCompilation` which gets written at the top level as well as a nested reference. Having the `"compilation": ` indirection in the second case seemed odd (willing to be convinced on this point though)